src/commands/classified/external.rs

```diff
@@ -50,158 +49,243 @@ impl Decoder for LinesCodec {
+        trace!(target: "nu::run::external", "set up stdout pipe");
+    }
 
-            format!("{} {}", command.name, itertools::join(args, " "))
-        });
+    trace!(target: "nu::run::external", "built process {:?}", process);
 
-        process = Exec::shell(itertools::join(commands, " && "))
-    } else {
-        process = Exec::cmd(&command.name);
-        for arg in command.args.iter() {
-            // Let's also replace ~ as we shell out
-            let arg = shellexpand::tilde_with_context(arg.deref(), || home_dir.as_ref());
-
-            let arg_chars: Vec<_> = arg.chars().collect();
-            if arg_chars.len() > 1 && arg_chars[0] == '"' && arg_chars[arg_chars.len() - 1] == '"' {
-                // quoted string
-                let new_arg: String = arg_chars[1..arg_chars.len() - 1].iter().collect();
-                process = process.arg(new_arg);
-            } else {
-                process = process.arg(arg.as_ref());
+    let popen = process.popen();
+    if let Ok(mut popen) = popen {
+        let _ = popen.detach();
+        if is_last {
+            while let None = popen.poll() {
+                std::thread::sleep(std::time::Duration::new(0, 100000000));
             }
+
+            let _ = popen.terminate();
+            Ok(InputStream::empty())
+        } else {
+            let stdout = popen.stdout.take().unwrap();
+            let file = futures::io::AllowStdIo::new(stdout);
+            let stream = Framed::new(file, LinesCodec {});
+            let stream = stream.map(move |line| line.unwrap().into_value(&name_tag));
+            let stream = stream.boxed() as BoxStream<'static, Value>;
+            Ok(stream.into())
         }
+    } else {
+        Err(ShellError::labeled_error(
+            "Command not found",
+            "command not found",
+            name_tag,
+        ))
     }
+}
 
-    process = process.cwd(context.shell_manager.path());
+async fn run_with_stdin(
+    command: ExternalCommand,
+    context: &mut Context,
+    mut input: InputStream,
+    is_last: bool,
+) -> Result<InputStream, ShellError> {
+    let name_tag = command.name_tag.clone();
+    let home_dir = dirs::home_dir();
 
-    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
+    let mut process = Exec::cmd(&command.name);
+    for arg in command.args.iter() {
+        // Let's also replace ~ as we shell out
+        let arg = shellexpand::tilde_with_context(arg.deref(), || home_dir.as_ref());
 
-    let mut process = match stream_next {
-        StreamNext::Last => process,
-        StreamNext::External | StreamNext::Internal => {
-            process.stdout(subprocess::Redirection::Pipe)
+        let arg_chars: Vec<_> = arg.chars().collect();
+        if (arg_chars.len() > 1 && arg_chars[0] == '"' && arg_chars[arg_chars.len() - 1] == '"')
+            || (arg_chars.len() > 1
+                && arg_chars[0] == '\''
+                && arg_chars[arg_chars.len() - 1] == '\'')
```

Comment:
I should have called out this small change too. It strips out the single quotes (e.g., `^echo 'abc'` will emit `abc` instead of `'abc'`)

Reply:
Sounds good